src/Microsoft.CodeQuality.Analyzers.Exp/Core/FlowAnalysis/Analysis/DisposeAnalysis/DisposeAnalysis.DisposeDataFlowOperationVisitor.cs

```diff
@@ -0,0 +1,302 @@
+                INamedTypeSymbol genericICollection,
+                ImmutableHashSet<INamedTypeSymbol> disposeOwnershipTransferLikelyTypes,
+                DisposeAbstractValueDomain valueDomain,
+                INamedTypeSymbol containingTypeSymbol,
+                DataFlowAnalysisResult<PointsToBlockAnalysisResult, PointsToAbstractValue> pointsToAnalysisResult,
+                DataFlowAnalysisResult<NullBlockAnalysisResult, NullAbstractValue> nullAnalysisResultOpt)
+                : base(valueDomain, containingTypeSymbol, nullAnalysisResultOpt: nullAnalysisResultOpt, pointsToAnalysisResultOpt: pointsToAnalysisResult)
+            {
+                Debug.Assert(iDisposable != null);
+                Debug.Assert(disposeOwnershipTransferLikelyTypes != null);
+                Debug.Assert(pointsToAnalysisResult != null);
+
+                _iDisposable = iDisposable;
+                _iCollection = iCollection;
+                _genericICollection = genericICollection;
+                _disposeOwnershipTransferLikelyTypes = disposeOwnershipTransferLikelyTypes;
+            }
+
+            protected override DisposeAbstractValue GetAbstractDefaultValue(ITypeSymbol type) => DisposeAbstractValue.NotDisposable;
+
+            protected override DisposeAbstractValue GetAbstractValue(AbstractLocation location) => CurrentAnalysisData.TryGetValue(location, out var value) ? value : ValueDomain.UnknownOrMayBeValue;
+
+            protected override void SetAbstractValue(AbstractLocation location, DisposeAbstractValue value)
+            {
+                Debug.Assert(location.LocationType.IsDisposable(_iDisposable));
+
+                CurrentAnalysisData[location] = value;
+            }
+
+            protected override void SetAbstractValue(PointsToAbstractValue instanceLocation, DisposeAbstractValue value)
+            {
+                foreach (var location in instanceLocation.Locations)
+                {
+                    SetAbstractValue(location, value);
+                }
+            }
+
+            protected override void ResetCurrentAnalysisData(DisposeAnalysisData newAnalysisDataOpt = null)
+            {
+                // Reset the current analysis data, while ensuring that we don't violate the monotonicity, i.e. we cannot remove any existing key from currentAnalysisData.
+                if (newAnalysisDataOpt == null)
+                {
+                    // Just set the values for existing keys to ValueDomain.UnknownOrMayBeValue.
+                    var keys = CurrentAnalysisData.Keys.ToImmutableArray();
+                    foreach (var key in keys)
+                    {
+                        SetAbstractValue(key, ValueDomain.UnknownOrMayBeValue);
+                    }
+                }
+                else
+                {
+                    // Merge the values from current and new analysis data.
+                    var keys = CurrentAnalysisData.Keys.Concat(newAnalysisDataOpt.Keys).ToImmutableArray();
```

Comment:
Should be a set, or we'll calculate merges multiple times for keys that exist in both datasets, which could potentially cause bugs.

Reply:
If we're guaranteed that these won't overlap, then I'd like to see assert for that.

---
In reply to: [168852927](https://github.com/dotnet/roslyn-analyzers/pull/1583#discussion_r168852927) [](ancestors = 168852927)

Reply:
Makes, sense. I will change this to create an immutable set